packages/app/src/cli/services/dev/graphiql/server.ts

```diff
@@ -87,87 +53,141 @@ export function setupGraphiQLServer({
       next()
     })
 
-  let session: Session | undefined
-
-  async function beginAuth(req: express.Request, res: express.Response) {
-    stdout.write('Initiating OAuth flow...')
-    await shopify.auth.begin({
-      shop: shopify.utils.sanitizeShop(storeFqdn, true)!,
-      callbackPath: `/${urlNamespaces.devTools}/graphiql/auth/callback`,
-      isOnline: false,
-      rawRequest: req,
-      rawResponse: res,
-    })
+  let _token: string | undefined
+  async function token(): Promise<string> {
+    if (!_token) {
+      // eslint-disable-next-line require-atomic-updates
+      _token = await refreshToken()
+    }
+    return _token
   }
 
-  async function isAuthorized(): Promise<boolean> {
-    if (!session) return false
-    const client = new shopify.clients.Graphql({session, apiVersion: LATEST_API_VERSION})
+  async function refreshToken(): Promise<string> {
     try {
-      await client.query({data: '{ shop { id } }'})
-      return true
-      // eslint-disable-next-line no-catch-all/no-catch-all
-    } catch (error) {
-      return (error as HttpResponseError)?.response?.code !== 401
+      outputDebug('refreshing token', stdout)
+      _token = undefined
+      const queryString = queryStringEncode({
+        client_id: apiKey,
+        client_secret: apiSecret,
+        grant_type: 'client_credentials',
+      })
+      const tokenResponse = await fetch(`https://${storeFqdn}/admin/oauth/access_token?${queryString}`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      })
+      const tokenJson = (await tokenResponse.json()) as {access_token: string}
+      return tokenJson.access_token
+    } catch (_error) {
+      throw new TokenRefreshError()
     }
   }
 
   app.get('/graphiql/ping', (_req, res) => {
     res.send('pong')
   })
 
-  // eslint-disable-next-line @typescript-eslint/no-misused-promises
-  app.get('/graphiql/auth/callback', async (req, res) => {
-    outputDebug('Handling /graphiql/auth/callback request', stdout)
-    // The library will automatically set the appropriate HTTP headers
-    const callback = await shopify.auth.callback({
-      rawRequest: req,
-      rawResponse: res,
-    })
-    session = callback.session
-    // You can now use callback.session to make API requests
-    res.redirect(`/${urlNamespaces.devTools}/graphiql`)
+  const stylePath = require.resolve('@shopify/cli-kit/assets/style.css')
+  app.get('/graphiql/simple.css', (_req, res) => {
+    res.sendFile(stylePath)
+  })
+
+  async function fetchApiVersionsWithTokenRefresh(): Promise<string[]> {
+    let apiVersions: string[]
+    try {
+      apiVersions = await supportedApiVersions({storeFqdn, token: await token()})
+      // eslint-disable-next-line no-catch-all/no-catch-all
+    } catch (err) {
+      // Retry once with a new token, in case the token expired or was revoked
+      await refreshToken()
```

Comment:
Feel free to pass on this, not essential

We have a few places now where we have a pattern of:

```
try {
  return doAThing(token)
} catch (err) {
  token = await refreshToken()
  return doAThing(token)
}
```

I wonder if we can provide a helper for this somewhere?

Reply:
Meaning, it's easy to build this structure incorrectly especially with async/await. Like so:

```js
// doAThing is async, this is wrong:

async function myFunc() {
  try {
    // Usually you can return a promise fro an async function,
    // but in this case if it fails there will be no retry!
    return doAThing(token)
  } catch (err) {
    token = await refreshToken()
    return doAThing(token)
  }
}

// should be:

async function myFunc() {
  let retval
  try {
    retval = doAThing(token)
  } catch (err) {
    token = await refreshToken()
    retval = doAThing(token)
  }
  return retval
}

```

So it would be good to have as a util. Sorry I was unclear before!

Reply:
Fine by me to tackle it in a separate thing